velox/functions/sparksql/aggregates/ModeAggregate.cpp

```diff
@@ -0,0 +1,573 @@
+
+  struct AccumulatorType {
+    using ValueMap = folly::F14FastMap<
+        T,
+        int64_t,
+        Hash,
+        EqualTo,
+        AlignedStlAllocator<std::pair<const T, int64_t>, 16>>;
+
+    // A map of T -> count.
+    ValueMap values;
+
+    explicit AccumulatorType(HashStringAllocator* allocator)
+        : values{AlignedStlAllocator<std::pair<const T, int64_t>, 16>(
+              allocator)} {}
+    static constexpr bool is_fixed_size_ = false;
+
+    void addInput(HashStringAllocator* /*allocator*/, arg_type<T> data) {
+      values[data]++;
+    }
+
+    void combine(
+        HashStringAllocator* /*allocator*/,
+        arg_type<IntermediateType> other) {
+      for (const auto& [key, count] : other) {
+        if (count.has_value()) {
+          values[key] += count.value();
+        }
+      }
+    }
+
+    bool writeIntermediateResult(out_type<IntermediateType>& out) {
+      auto size = values.size();
+      out.reserve(size);
+      for (const auto& [value, count] : values) {
+        out.emplace(value, count);
+      }
+      return true;
+    }
+
+    bool writeFinalResult(out_type<OutputType>& out) {
+      auto iterator = std::max_element(
+          std::begin(values),
+          std::end(values),
+          [](const std::pair<T, int64_t>& p1, const std::pair<T, int64_t>& p2) {
+            return p1.second < p2.second;
+          });
+      out = iterator->first;
+      return true;
+    }
+
+    void destroy(HashStringAllocator* /*allocator*/) {
+      using TValues = decltype(values);
+      values.~TValues();
+    }
+  };
+};
+
+class StringModeAggregate {
+ public:
+  using InputType = Row<Varchar>;
+
+  using IntermediateType = Map<Varchar, int64_t>;
+
+  using OutputType = Varchar;
+
+  struct AccumulatorType {
+    using ValueMap = folly::F14FastMap<
+        StringView,
+        int64_t,
+        std::hash<StringView>,
+        std::equal_to<StringView>,
+        AlignedStlAllocator<std::pair<const StringView, int64_t>, 16>>;
+
+    // A map of unique StringViews pointing to storage managed by 'strings'.
+    ValueMap values;
+
+    // Stores unique non-null non-inline strings.
+    Strings strings;
+
+    explicit AccumulatorType(HashStringAllocator* allocator)
+        : values{AlignedStlAllocator<std::pair<const StringView, int64_t>, 16>(
+              allocator)} {}
+    static constexpr bool is_fixed_size_ = false;
+
+    void addInput(HashStringAllocator* allocator, arg_type<Varchar> data) {
+      values[store(data, allocator)]++;
+    }
+
+    void combine(
+        HashStringAllocator* allocator,
+        arg_type<IntermediateType> other) {
+      for (const auto& [key, count] : other) {
+        if (count.has_value()) {
+          values[store(key, allocator)] += count.value();
+        }
+      }
+    }
+
+    bool writeIntermediateResult(out_type<IntermediateType>& out) {
+      auto size = values.size();
+      out.reserve(size);
+      for (const auto& [value, count] : values) {
```

Comment:
Can we use a same struct for `values` and `out`, then we can use std::move() for them, not process each value.
I'm not sure if this can make sense.

Reply:
I notice `HistogramAggregate` does not have this function, for this performance issue?

Reply:
They are different types, immediate type is MAP<T, BIGINT>, another is F14FastMap. We can't call move function on them.

> I notice HistogramAggregate does not have this function, for this performance issue?

In `HistogramAggregate`, both immediate type and result type is MAP<T, BIGINT>, so they resuse same function to extract result values and immediate values. This is not related to any perf issues.

Reply:
`HistogramAggregate`  intermediate type is also `ValueMap` https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/HistogramAggregate.cpp#L42

Reply:
I think intermediate type should be some Velox types not the ValueMap. Sorry, I don't get what's your point here. 
https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/aggregates/HistogramAggregate.cpp#L493